### PR TITLE
Add Interactor guide - AddBook

### PR DESCRIPTION
**Pull request by cllns**
Looking for feedback on this. It's still a **work-in-progress**

I'm concerned this gets a _little_ too into the weeds about specifics of authentication, but I feel like it'd be irresponsible to not talk about security concerns here.

Planning on adding sections about:
- Using Interactors in Actions - including testing
- Validations - `.valid?`
- Controlling flow -  the difference between `error`, `error!`, `fail!`
- Rationale - when to use Interactors

Do you think those three sections belong in the guide? Is there anything I'm leaving out? (Maybe composing them?)

I just wrote it and haven't thoroughly edited it yet, so I'm not looking for typo/spelling/grammar fixes yet. It's more big picture right now.

Addresses #215

### CODE REVIEW

**Comment by sandelius** [comment]
I definitely do since we're pushing for this and it's not super common for newer developer.

File: source/guides/architecture/interactors.md [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+      string :password_digest
+
+      column :created_at, DateTime, null: false
+      column :updated_at, DateTime, null: false
+    end
+  end
+end
+```
+
+We'll add three columns:
```
Comment (atzorvas) [inline source/guides/architecture/interactors.md]: **two** columns

File: source/guides/architecture/interactors.md [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+
+Likewise, we need to fill out a new spec: `spec/bookshelf/interactors/sign_up_user_spec.rb`
+
+```ruby
+require 'spec_helper'
+
+describe SignUpUser do
+  it "succeeds" do
+    interactor.call.success?.must_equal true
```
Comment (atzorvas) [inline source/guides/architecture/interactors.md]: does this need ```let(:interactor) { SignUpUser.new }``` ?

File: source/guides/architecture/interactors.md [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+  end
+end
+```
+
+These are the only two public methods this class has:
+`initialize`, to set-up the data, and
+`call` to actually perform the operation(s).
+
+
+Likewise, we need to fill out a new spec: `spec/bookshelf/interactors/sign_up_user_spec.rb`
```
Comment (atzorvas) [inline source/guides/architecture/interactors.md]: bookshelf guide is test-first. so the test could be created before the actual class.

File: source/guides/architecture/interactors.md [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+We just need to clean up the database before we run each test.
+
+Edit `spec/bookshelf/interactors/sign_up_user_spec.rb`:
+```ruby
+require 'spec_helper'
+
+describe SignUpUser do
+  let(:interactor) { SignUpUser.new(email: "test@example.com") }
+
+  before { UserRepository.new.clear }
```
Comment (atzorvas) [inline source/guides/architecture/interactors.md]: I think the following is more in-sync with the rest of the bookshelf guide

```ruby
let(:repository) { UserRepository.new }

before do
  repository.clear
end
```

File: source/guides/architecture/interactors.md [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+BCrypt includes a salt (so you don't have to worry about that)
+and includes key-stretching, which makes them harder to crack.
+
+Let's ensure the `User` has a password_digest:
+```ruby
+describe SignUpUser do
+  # ...
+
+  it "creates a User with password_digest" do
+    result.user.password_digest.length.must_equal 60
```
Comment (atzorvas) [inline source/guides/architecture/interactors.md]: ```interactor.user``` instead of ```result.user```

**Comment by cllns** [comment]
Thanks @atzorvas! I fixed all of them except the repository tip. I'll switch that syntax over if we end up needing the `repository` for any other commands :)

File: source/guides/architecture/interactors.md [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+Instead, attackers would only get the hashed results.
+
+Note: this approach is still vulnerable to
+[Rainbow tables](https://en.wikipedia.org/wiki/Rainbow_table).
+
+In order to combat this, we'll use a hashing scheme that employs a
+[`salt`](https://en.wikipedia.org/wiki/Salt_(cryptography)).
+
+# Preparing
```
Comment (davydovanton) [inline source/guides/architecture/interactors.md]: maybe is better to add only link to `bookshelf` app?

File: source/guides/architecture/interactors.md [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+```bash
+git clone git@github.com:hanami/bookshelf.git hanami-bookshelf
+cd hanami-bookshelf
+bundle install
+HANAMI_ENV=development hanami db prepare
+HANAMI_ENV=test hanami db prepare
+bundle exec rake # All the tests should pass
+```
+
+# Adding User model
```
Comment (davydovanton) [inline source/guides/architecture/interactors.md]: I think it's not a relevant to `Interactors` information WDYT? 
I understand that you want to explain interactors as more as possible but user can find all this information on other parts of guides.

File: source/guides/architecture/interactors.md [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+- `password_digest`, this will be the user's hashed password
+
+
+Let's create that table:
+
+```bash
+bundle exec hanami db prepare
+```
+
+# Interactor
```
Comment (davydovanton) [inline source/guides/architecture/interactors.md]: @jodosha just a idea:
maybe we introduce new cli command `hanami generate interactor <interactors_name>` ?

File: source/guides/architecture/interactors.md [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+and write a new spec `spec/bookshelf/interactors/sign_up_user_spec.rb`:
+
+```ruby
+require 'spec_helper'
+
+describe SignUpUser do
+  let(:interactor) { SignUpUser.new(email: "test@example.com") }
+
+  it "succeeds" do
+    interactor.call.success?.must_equal true
```
Comment (davydovanton) [inline source/guides/architecture/interactors.md]: rspec style for tests?

**Comment by davydovanton** [comment]
@cllns any progress here? Can I help you to complete this guide?

File: source/guides/architecture/interactors.md [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+
+Instead, we recommend being **explicit over implicit**.
+
+An interactor is an object that represents a specific *use-case*.
+
+They let each class have a single responsibility.
+Interactors' single responsibility is to call other behavior.
+
+We provide `Hanami::Interactor` as a module,
+so you can start with a Plain Old Ruby Object (no superclass),
```
Comment (beauby) [inline source/guides/architecture/interactors.md]: Moreover, the `Object` class _does_ have a superclass (namely `BasicObject`).

**Comment by cllns** [comment]
Thanks @beauby! Still working on this, and I'll integrate your suggestions.

**Comment by cllns** [comment]
Note: I was originally going to make this interactor `SignUpUser` but now I'm working on an `AddBook` interactor, which sends an email notification. This removes the need to describe some basics of authentication, which is outside the scope of this guide.

**Comment by cllns** [comment]
Hi everyone, just pushed a new version of this guide. It's in RSpec (already started converting the Getting Started guide over too). Please review when you get a chance @hanami/core @beauby

I'm sure there are typos and redundant language. Haven't had a chance to clean up the language yet. Also, I want to add more examples later (error handling, valid?, injecting into an action)... but perfect has been the enemy of good for too long already :)

**Comment by jodosha** [comment]
@cllns Please hold on with the merge of this, until we release 1.1.

**Comment by cllns** [comment]
Merged via https://github.com/hanami/hanami.github.io/commit/495bf297c8d65b47b067a1f0965fe1c280cf3e1c

**Pull request closed without merging**